CI, Build and Test updates

matrei opened the pull request:
- Stabilize flaky `CustomAutoTimestampSpec`
  - Address intermittent CI failures possibly caused by date values not incrementing as expected.
  - Enhance assertions by comparing `.time` values to reveal the actual millisecond differences.
  - Remove annotations that were added as a try to remedy the flakiness, to determine if they are warranted.
  - Omit cleanup of `insertOverwrite`, to determine if it is necessary for test stability.
  - Do one comparison per statement in `then` blocks.
- Fix `GroovyPageUnitTestMixinWithCustomViewDirSpec` on Windows
  - The test was failing on Windows due to line ending differences affecting string comparisons.
- Apply `asset-pipeline` Gradle plugin in some test applications with assets
  - This puts `org.graalvm.js:js-community` on the `runtimeClasspath` and avoids `ClassNotFoundException: org.graalvm.polyglot.Context`.
- Re-enable `TckTestSuite` on Windows as it does not seem to fail
- Update `gradle.yml`
- Fix an incorrect log pattern in functional test applications
- Update deprecated `converterClass` attributes in `logback.xml`
- Fix develocity authentication in `grails-gradle`
- Switch some Gradle log statements from `lifecycle` to `info` to clean up the log output a bit.
- Update Gradle Plugin e2e tests to use Groovy 2.0.25.
- Add `mavenCentral` repository to Gradle Plugin e2e tests.
- Switch to using `@PendingFeatureIf` for tests that only fail in `CI`.

matrei committed:
test: stabilize flaky `CustomAutoTimestampSpec`

- Address intermittent CI failures possibly caused by date
values not incrementing as expected.
- Enhance assertions by comparing `.time` values to reveal
the actual millisecond differences.
- Remove annotations that were added as a try to remedy the
flakiness, to determine if they are warranted.
- Omit cleanup of `insertOverwrite`, to determine if it
is necessary for test stability.
- Do one comparison per statement in `then` blocks.

matrei committed:
test: fix `GroovyPageUnitTestMixinWithCustomViewDirSpec` on Windows

The test was failing on Windows due to line-ending differences affecting
string comparisons.

matrei committed:
test: apply `asset-pipeline` plugin

- Apply the `asset-pipeline` Gradle plugin to functional test projects where
it was missing.
- This adds `org.graalvm.js:js-community (required by `asset-pipeline-grails`)
to the `runtimeClasspath`.
- Without this, we get `ClassNotFoundException: org.graalvm.polyglot.Context`.

matrei committed:
test: re-enable `TckTestSuite` on Windows

Attempting to re-enable the test on Windows, as it seems to pass
in local environments.

matrei committed:
ci: update `gradle.yml`

- Clarify comment why `GITHUB_TOKEN` is added as an environment variable.
- Align lists to use space padding for readability.
- Align `.gradlew` linebreaks.
- Align `setup-java` params to use alphabetical ascending order.
- Remove unused step ids.
- Use `bootJar` and `check` task instead of `build` in functional test
jobs as we don't have to do a full build there (docs etc.).
- Align on adding `--continue` and `--stacktrace` for test jobs.
- Align on `ubuntu-24.04` instead of `ubuntu-latest` for non-os-matrixed
jobs (for stability).

matrei committed:
test: fix an incorrect log pattern in functional test apps

matrei committed:
test: update deprecated `logback.xml` `converterClass` attribute

Removes log output:
[converterClass] attribute is deprecated and replaced by [class]

matrei committed:
build: make handling of `skip*` and `only*` more declarative

matrei committed:
build: fix develocity authentication

matrei committed:
build: switch from `lifecycle` to `info` logger

To make the build a little less noisy.

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -212,20 +213,18 @@ jobs:
+    runs-on: ubuntu-24.04
     permissions:
-      contents: write
+      contents: write # TODO: Comment why this is needed
```
This isn't needed because we will commit to a different repo

matrei replied:
OK, so I can remove it?

I also see some `contents: read  #  to fetch code (actions/checkout)`
These should not be necessary, should they?

jdaugherty replied:
read is still needed to checkout, write isn't

matrei replied:
I think read is the default, so it is implied if you don't specify anything: https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

jdaugherty replied:
Le'ts give it a try then!

jdaugherty commented on grails-datamapping-core-test/src/test/groovy/org/grails/datastore/gorm/TckTestSuite.groovy:
```diff
@@ -21,14 +21,12 @@ package org.grails.datastore.gorm
  *
  * @author graemerocher
  */
-@IgnoreIf({ os.windows }) // Fails on windows even though the NotInListSpec is in the TCK and runs
```
I ran this multiple times and it kept failing in CI.  James F couldn't reproduce locally though.  We never could explain why it was failing.  We don't need it for CI, so I suggest we just disable if GITHUB_REF is present

matrei replied:
It seems it is no longer failing in CI, so we should not need to disable it at all.

jdaugherty replied:
We can try it, but in my deterministic ordering PR, I had it fail twice randomly.

jdaugherty commented on grails-datamapping-core-test/src/test/groovy/org/grails/datastore/gorm/CustomAutoTimestampSpec.groovy:
```diff
@@ -94,6 +93,7 @@ class CustomAutoTimestampSpec extends GormDatastoreSpec {
         def now = new Date()
         r.created = new Date(now.time)
         r.modified = r.created
+        sleep(50) // give the save a chance to set a different time
```
Can't this just be sleep(1)?

matrei replied:
Probably, I'll try it out.

jdaugherty commented on grails-datamapping-core-test/src/test/groovy/org/grails/datastore/gorm/CustomAutoTimestampSpec.groovy:
```diff
@@ -22,15 +22,9 @@ import grails.gorm.annotation.AutoTimestamp
 
 import static grails.gorm.annotation.AutoTimestamp.EventType.CREATED
 
-@Retry // this test is flaky on CI due to https://github.com/apache/grails-data-mapping/issues/1877
```
I think this is still failing because there is a shared, static variable thats being updated separately by another test.  The comment should stay

matrei replied:
Is it still failing? Where?

jdaugherty replied:
it fails in parallel tests; i have had it fail locally sporadically and it seems to consistently fail on the windows 23 build

matrei replied:
>  it seems to consistently fail on the windows 23 build

I have not seen `CustomAutoTimestampSpec` fail even once on Windows/23 after the changes in this branch: https://github.com/apache/grails-core/actions/runs/14657676325/job/41135500844?pr=14668

Do I have to change anything to the test setup to trigger the failure?

jdaugherty replied:
Let's try it with your changes and then if it fails, I'll ignore it again?

jdaugherty commented on grails-datamapping-core-test/src/test/groovy/org/grails/datastore/gorm/CustomAutoTimestampSpec.groovy:
```diff
@@ -41,19 +35,21 @@ class CustomAutoTimestampSpec extends GormDatastoreSpec {
         then:"the custom lastUpdated and dateCreated are set"
-            r.modified != null && r.modified < new Date()
-            r.created != null && r.created < new Date()
+            r.modified != null
```
You could add a sleep(1) and then restore these time checks

matrei replied:
I don't think the time checks are necessary to fulfill the contract of what is tested.
That is why I removed it. Do you think I should add it back?

jdaugherty replied:
I think we should

jdaugherty commented on grails-datamapping-core-test/src/test/groovy/org/grails/datastore/gorm/CustomAutoTimestampSpec.groovy:
```diff
@@ -106,16 +106,15 @@ class CustomAutoTimestampSpec extends GormDatastoreSpec {
         Date previousCreated = r.created
         Date previousModified = r.modified
         r.name = "Test 2"
+        sleep(50) // give the save a chance to set a different time
```
Doesn't this just need to be sleep(1)?

jdaugherty commented on grails-datamapping-core-test/src/test/groovy/org/grails/datastore/gorm/CustomAutoTimestampSpec.groovy:
```diff
@@ -63,24 +59,27 @@ class CustomAutoTimestampSpec extends GormDatastoreSpec {
         Date previousCreated = r.created
         Date previousModified = r.modified
         r.name = "Test 2"
+        sleep(50) // give the save a chance to set a different time
```
sleep(1)?

jdaugherty commented on grails-datamapping-core-test/src/test/groovy/org/grails/datastore/gorm/CustomAutoTimestampSpec.groovy:
```diff
@@ -63,24 +59,27 @@ class CustomAutoTimestampSpec extends GormDatastoreSpec {
         def now = new Date()
         r.created = new Date(now.time)
         r.modified = r.created
+        sleep(50) // give the save a chance to set a different time
```
sleep(1)?

jdaugherty commented on grails-datamapping-core-test/src/test/groovy/org/grails/datastore/gorm/CustomAutoTimestampSpec.groovy:
```diff
@@ -41,19 +35,21 @@ class CustomAutoTimestampSpec extends GormDatastoreSpec {
             Date previousCreated = r.created
             Date previousModified = r.modified
             r.name = "Test 2"
+            sleep(50) // give the save a chance to set a different time
```
sleep(1)

matrei committed:
test: shorten the sleep timeout in `CustomAutoTimestampSpec`

matrei committed:
test: add `mavenCentral()` to Gradle e2e tests

This prevents timeouts when `repo.grails.org` is proxying.

matrei committed:
test: update Gradle e2e tests to Groovy version 4.0.25

jdaugherty commented on grails-gradle/plugins/src/e2eTest/resources/publish-projects/legacy-apply/child-project-with-unrelated-parent/build.gradle:
```diff
@@ -20,6 +20,7 @@
 buildscript {
     repositories {
         maven { url "${System.getenv('LOCAL_MAVEN_PATH')}" }
+        mavenCentral()
```
We should remove the grails repo & the apache repo.  They shouldnt' be needed.  the purpose of these tests were to have a minimum build file (so just include groovy as a minimum dependency), and the gradle plugin should be substituted by the gradle test kit.

matrei replied:
It seems the tests fail to find all dependencies if the repositories are removed.

jdaugherty replied:
thats unfortunate; thank you for trying

jdaugherty commented on grails-gradle/plugins/src/e2eTest/resources/publish-projects/legacy-apply/child-project-with-unrelated-parent/build.gradle:
```diff
@@ -20,6 +20,7 @@
 buildscript {
     repositories {
         maven { url "${System.getenv('LOCAL_MAVEN_PATH')}" }
+        mavenCentral()
```
We should remove the grails repo & the apache repo.  They shouldnt' be needed.  the purpose of these tests were to have a minimum build file (so just include groovy as a minimum dependency), and the gradle plugin should be substituted by the gradle test kit.

jdaugherty commented on grails-test-suite-uber/src/test/groovy/grails/test/mixin/GroovyPageUnitTestMixinWithCustomViewDirSpec.groovy:
```diff
@@ -35,6 +35,6 @@ class GroovyPageUnitTestMixinWithCustomViewDirSpec extends Specification impleme
         
         then:
-        result == '\nthis is a custom template'
+        result.contains('this is a custom template')
```
Should we just call trim() and compare the resulting string instead?

matrei replied:
OK, makes sense!

matrei committed:
ci: remove all permission definitions from `gradle.yml`

The default permissions should suffice for what we are doing.

matrei committed:
ci: align on using unquoted java versions

matrei committed:
feedback(test): use `trim()` in `GroovyPageUnitTestMixinWithCustomViewDirSpec`

matrei committed:
feedback(test): restore date comparisons in `CustomAutoTimestampSpec`

matrei committed:
test: try removing `@Ignore` on `FormTagLibTests`

matrei committed:
test: use `pageSource.contains()` in `InheritedConfigSpec`

This will show the page source output on test failures.

matrei committed:
test: refactor `FormTagLibTests`

matrei committed:
test: add `@PendingFeatureIf` to `FormTagLibTest`

`testFormActionSubmitWithController` fails for some reason only in CI.
It seems that `createLink` from `ApplicationTagLib` returns an empty string.
